lib/Toolchain/Compiler.cpp

```diff
@@ -38,3 +45,76 @@ OwningBinary<ObjectFile> Compiler::compileModule(Module *module, TargetMachine &
 
   return objectFile;
 }
+
+std::string Compiler::compileBitcode(const Bitcode &bitcode) {
+  std::string error;
+  llvm::raw_string_ostream stream(error);
+  if (llvm::verifyModule(*bitcode.getModule(), &stream)) {
+    stream.flush();
+    diagnostics.error(error);
+  }
+
+  const std::string &targetTriple = bitcode.getModule()->getTargetTriple();
+  const llvm::Target *target = llvm::TargetRegistry::lookupTarget(targetTriple, error);
+  if (!target) {
+    diagnostics.error("Cannot lookup target: "s + error + '\n');
+    return std::string();
+  }
+
+  auto CPU = "generic";
+  auto features = "";
+  llvm::TargetOptions opt;
+  llvm::Optional<llvm::Reloc::Model> relocationModel;
+  llvm::TargetMachine *targetMachine =
+      target->createTargetMachine(targetTriple, CPU, features, opt, relocationModel);
+
+  llvm::Twine prefix("mull");
+  llvm::SmallString<128> resultPath;
+
+  if (std::error_code err = llvm::sys::fs::createTemporaryFile(prefix, "o", resultPath)) {
+    diagnostics.error("Cannot create temporary file"s + err.message());
+    return std::string();
+  }
+  std::error_code errorCode;
+  llvm::raw_fd_ostream dest(resultPath, errorCode, llvm::sys::fs::OpenFlags::F_None);
+
+  if (errorCode) {
+    diagnostics.error("Could not open file: "s + errorCode.message());
+    return std::string();
+  }
+
+  llvm::legacy::PassManager pass;
+  if (llvm_compat::addPassesToEmitObjectFile(targetMachine, pass, dest)) {
+    diagnostics.error("TargetMachine can not emit object file");
+    return std::string();
+  }
+
+  for (llvm::Function &function : *bitcode.getModule()) {
+    if (function.isDSOLocal()) {
+      function.setDSOLocal(false);
+      function.setLinkage(llvm::GlobalValue::InternalLinkage);
+    }
+  }
+
+  if (configuration.debugEnabled) {
+    llvm::SmallString<128> bcPath;
+
+    if (std::error_code err = llvm::sys::fs::createTemporaryFile(prefix, "bc", bcPath)) {
+      diagnostics.error("Cannot create temporary file"s + err.message());
+      return std::string();
+    }
+    llvm::raw_fd_ostream bcStream(bcPath, errorCode, llvm::sys::fs::OpenFlags::F_None);
+    if (errorCode) {
+      diagnostics.warning("Could not open temp bc file: "s + errorCode.message());
+      return std::string();
+    }
+    llvm_compat::writeBitcodeToFile(*bitcode.getModule(), bcStream);
+    diagnostics.debug("Emitted object file: "s + resultPath.c_str() + " for bitcode file: "s +
+                      bcPath.c_str());
+  }
+
+  pass.run(*bitcode.getModule());
+  dest.flush();
+
+  return resultPath.str().str();
+}
```

Comment:
This piece uses LLVM's API to generate an object file. It may make sense to actually just rely on clang/clang++ in case we have a compilation database. Otherwise, we can wrongly emit 'slightly' incorrect/unexpected object files due to different defaults.